docs/plugin_developers_guide/07_rules_plugins.rst

```diff
@@ -0,0 +1,432 @@
+notified to that other system or device.
+
+
+Notification delivery plugins may be written in C or C++ and have a very
+simple interface. The plugin mechanism and a subset of the API is common
+between all types of plugins including notification rules.  This documentation
+is based on the |code|. The |average| calculates a moving average of the values
+sent to the notification instance and will trigger the notification to be sent
+when the value of a datapoint varies from the calculated average by more than
+a specified percentage.
+
+Configuration
+-------------
+
+Notification Rule  plugins use the same configuration mechanism as the rest of
+Fledge, using a JSON document to describe the configuration parameters. As
+with any other plugin the structure is defined by the plugin and retrieve
+by the *plugin_info* entry point. This is then matched with the database
+content to pass the configured values to the *plugin_init* entry point.
+
+Notification Rule Plugin API
+----------------------------
+
+The notification rule plugin API consists of a small number of C
+function entry points, these are called in a strict order and based on
+the same set of common API entry points for all Fledge plugins.
+
+Plugin Information
+~~~~~~~~~~~~~~~~~~
+
+The *plugin_info* entry point is the first entry point that is called
+in a notification rule plugin and returns the plugin information
+structure. This is the exact same call that every Fledge plugin
+must support and is used to determine the type of the plugin and the
+configuration category defaults for the plugin.
+
+A typical implementation of *plugin_info* would merely return a pointer
+to a static PLUGIN_INFORMATION structure.
+
+.. code-block:: C
+
+
+   PLUGIN_INFORMATION *plugin_info()
+   {
+        return &info;
+   }
+
+Plugin Initialise
+~~~~~~~~~~~~~~~~~
+
+The second call that is made to the plugin is the *plugin_init* call, that is used to retrieve a handle on the plugin instance and to configure the plugin.
+
+.. code-block:: C
+
+   PLUGIN_HANDLE plugin_init(ConfigCategory* config)
+   {
+           AverageRule *average = new AverageRule(config);
+           average->configure(config);
+           return (PLUGIN_HANDLE)average;
+   }
+
+
+The *config* parameter is the configuration category with the user supplied
+values inserted, these values are used to configure the behavior of the
+plugin. In the case of our moving average example we use this to call the constructor
+of our Average class and then call the configure method of that class.
+
+.. code-block:: C
+
+    /**
+     * Average rule constructor
+     *      
+     * Call parent class BuiltinRule constructor
+     */     
+    AverageRule::AverageRule() : BuiltinRule()
+    {       
+    }
+
+.. note::
+
+    We call the base class *BuiltinRule* as part of the construction of a
+    notification rule. This does some common initialisation required for all
+    notification rules.
```

Comment:
The base class *BuiltinRule* is invoked during the creation of a notification rule. This process performs essential initialization needed for all notification rules.